### DISCUSSION TITLE
SQLSTATE[23000]: Integrity constraint violation: 1062 Duplicate entry 'xxx' for key 'craft_freeform_submissions.freeform_submissions_incrementalId_unq_idx'

### DISCUSSION THREAD

**Discussion by nitriques** [solspace/craft-freeform#965] [Questions & Answers] [answered]
We do run a high load website that handles a lot of form submissions, on a daily basis.
The logs are filled with those kind of errors (with a different value each time).

**Steps to reproduce**

1. Create and launch a website with 15k form submissions per day (?)
2. Get the race condition to generate 2 same incremental ids.
3. Check the logs?

**Expected behavior**

Not have the race condition, so form submissions can be saved.

**Craft & Plugin Info (please complete the following information):**

- Craft Version: 4.5.12
- Freeform Version: 4.1.13
- Freeform Edition: Pro
- Fresh Install or Upgrade: Upgrade from 3.x

**Additional context**

I do not see why we need to keep track of this 'incremental' id.
This looks like the process of getting a new value then pushing it into the DB is not _atomic_, which would cause a race condition (2 requests getting the same id).

Can I at least DELETE the unique index ? The primary should be the incremental id no?
Where is this info used ?

Thanks a lot.

EDIT: Add stack trace

```
yii\db\IntegrityException: SQLSTATE[23000]: Integrity constraint violation: 1062 Duplicate entry '1886666' for key 'craft_freeform_submissions.freeform_submissions_incrementalId_unq_idx'
The SQL being executed was: INSERT INTO `craft_freeform_submissions` (`userId`, `formId`, `statusId`, `incrementalId`, `token`, `ip`, `isSpam`, `id`, `dateCreated`, `dateUpdated`, `uid`) VALUES (NULL, 111, 2, 1886666, 'clipped', '174.cc.vv.vv', 0, 2132428, '2024-01-11 11:35:45', '2024-01-11 11:35:45', '4c78e100-735d-vvvvv-848d-d0df906533b3')

#17 /home/vendor/yiisoft/yii2/db/Schema.php(676): yii\db\Schema::convertException
#16 /home/vendor/yiisoft/yii2/db/Command.php(1307): yii\db\Command::internalExecute
#15 /home/vendor/yiisoft/yii2/db/Command.php(1102): yii\db\Command::execute
#14 /home/vendor/solspace/craft-freeform/packages/plugin/src/Elements/Submission.php(452): Solspace\Freeform\Elements\Submission::afterSave
#13 /home/vendor/craftcms/cms/src/services/Elements.php(3382): craft\services\Elements::_saveElementInternal
#12 /home/vendor/craftcms/cms/src/services/Elements.php(1096): craft\services\Elements::saveElement
#11 /home/vendor/solspace/craft-freeform/packages/plugin/src/Services/SubmissionsService.php(203): Solspace\Freeform\Services\SubmissionsService::storeSubmission
#10 /home/vendor/solspace/craft-freeform/packages/plugin/src/Services/SubmissionsService.php(171): Solspace\Freeform\Services\SubmissionsService::handleSubmission
#9 /home/vendor/solspace/craft-freeform/packages/plugin/src/Controllers/SubmitController.php(64): Solspace\Freeform\Controllers\SubmitController::actionIndex
#8 [internal](0): call_user_func_array
#7 /home/vendor/yiisoft/yii2/base/InlineAction.php(57): yii\base\InlineAction::runWithParams
#6 /home/vendor/yiisoft/yii2/base/Controller.php(178): yii\base\Controller::runAction
#5 /home/vendor/yiisoft/yii2/base/Module.php(552): yii\base\Module::runAction
#4 /home/vendor/craftcms/cms/src/web/Application.php(305): craft\web\Application::runAction
#3 /home/vendor/craftcms/cms/src/web/Application.php(606): craft\web\Application::_processActionRequest
#2 /home/vendor/craftcms/cms/src/web/Application.php(284): craft\web\Application::handleRequest
#1 /home/vendor/yiisoft/yii2/base/Application.php(384): yii\base\Application::run
#0 /index.php(21): null
```

**Comment by kjmartens** [comment, marked as answer]
Sorry for the trouble you're experiencing, @nitriques.

The purpose of this incremental ID feature is to allow some customers to keep track of submission counts unique to Freeform (vs shared with all Craft Elements). It was a feature that a number of customers requested.

We will look into improving this in the near future. In the meantime, you should be able to delete the unique index without any consequence. 🙂

**Reply by nitriques** [reply]
Ok great ! Can you please confirm that having duplicates in that column will not crash the cp ?
By default, does freeform use this field ?

**Reply by kjmartens** [reply]
Sorry for the delay @nitriques. I didn't see your response until now. It will not crash Freeform. It's just used to show a unique number specific to Freeform for those who need it (sort of an edge-case feature), but Freeform doesn't rely on this anywhere. 🙂

**Reply by nitriques** [reply]
@kjmartens perfect, thanks !
